src/core/manager.c

```diff
@@ -1828,17 +1829,19 @@ static int manager_dispatch_signal_fd(sd_event_source *source, int fd, uint32_t
         }
 
         for (;;) {
-                n = read(m->signal_fd, &sfsi, sizeof(sfsi));
-                if (n != sizeof(sfsi)) {
-
-                        if (n >= 0)
-                                return -EIO;
-
-                        if (errno == EINTR || errno == EAGAIN)
+                ttl = 0;
+                do {
+                        n = read(m->signal_fd, buf + ttl, length - ttl);
+                        if (n < 0 && errno == EAGAIN) {
                                 break;
-
-                        return -errno;
-                }
+                        } else if (n < 0 && errno != EINTR) {
+                                log_error_errno(errno, "Failed to read bytes from signal file descriptor (%d): %m", errno); return -errno;
+                        } else if (n == 0) {
+                            log_error("Signal file descriptor received client disconnect, with %ld bytes discarded.", ttl); return -EIO; }
```

Comment:
What does "received client disconnect" mean? Who is the client? Disconnect from what?

http://man7.org/linux/man-pages/man2/signalfd.2.html
>  If  one  or  more of the signals specified in mask is pending for the process, then the buffer supplied to read(2) is used to return one or more signalfd_siginfo structures (see below) that describe the signals.  The read(2) returns information for as many signals as are pending and will fit in the supplied buffer.

Sorry, I don't really understand. Which problem are you trying to fix? Can you provide an example?